### DISCUSSION TITLE
add the ability to easily add a gpu to container for Memories and other apps that support hardware acceleration.

### DISCUSSION THREAD

**Discussion by relink2013** [nextcloud/all-in-one#1525] [Ideas] [answered]
It would be a very welcome addition if a GPU could easily be attached to the container to allow hardware transcoding. 

There are several apps in Nextcloud that can benefit from GPU acceleration, both Intel Quick Sync and Nvidia NVENC. Numerous other docker images have the ability to attach a GPU to use for this purpose, Plex, Jellyfin, etc...

Currently there doesn't seem to be a reliable way to accomplish this. Even after adding the drivers, and attaching the device to the `nextcloud-aio-nextcloud` it can be seen, but not used.

**Comment by szaimen** [comment]
> and attaching the device to the `nextcloud-aio-nextcloud`

Oh really? How did you do this?

**Reply by relink2013** [reply]
I installed the drivers by adding them to this variable in the compose file; 
`- NEXTCLOUD_ADDITIONAL_APKS=imagemagick nodejs npm ghostscript procps libbz2 stdbuf inotify bash ffmpeg libva-utils libva-vdpau-driver libva-intel-driver intel-media-driver mesa-va-gallium `

Then using portainer I edited the `nextcloud-aio-nextcloud` container and added the devices:
`/dev/dri/renderD128`
&
`/dev/dri/`
I tried multiple combinations of the two and got the same result every time. 

With that done, now when I run `occ memories:video-setup` I enable trasncoding, and now it says that Quick Sync is available and asks if I would like to use it. I answer yes, but when I go to play a video I still get "Transcode Failed" and the error log mentions a permission error.

**Reply by szaimen** [reply]
I see. The permission issue you can likely fix with this of https://github.com/pulsejet/memories/wiki/QSV-Transcoding#docker-installations
![Screenshot_20221216_194830_Brave](https://user-images.githubusercontent.com/42591237/208168054-7061723e-5dab-4d2a-a9be-5637ec8f44b6.png)

**Reply by szaimen** [reply]
Btw, bash and ffmpeg is already included in the image by default.

**Reply by szaimen** [reply]
Another btw, stdbuff is not a alpine package. What you are looking for is the coreutils package. See https://github.com/nextcloud/all-in-one/discussions/1531

**Reply by relink2013** [reply]
Thank you for the correction. I'm not that familiar with Alpine yet.

**Comment by szaimen** [comment]
I'll add the option to allow to add the dri device to the nextcloud container permanently at some point in the future.

**Reply by relink2013** [reply]
Until this gets "officially" added, is there any way to currently get it working at all that doesn't require rebuilding the image? I've decided to switch to a Quadro P400 instead of Quick Sync, and I'm trying to avoid installing Nextcloud in a VM just for this one thing.

**Reply by szaimen** [reply]
I fear NVENC will not possible anyway (see https://github.com/nextcloud/all-in-one/discussions/1536) but Intel Quick Sync will be.

**Reply by szaimen** [reply]
@Zoey2936 do you know if there is a docker image maybe which is compatible with alpine and comes with nvenc compatibility from which we maybe could get the ffmpeg binary?

**Reply by szaimen** [reply]
lets discuss this in https://github.com/nextcloud/all-in-one/discussions/1536

**Comment by szaimen** [comment]
This is now released with v4.1.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by relink2013** [reply]
This is Intel only right? I have been doing some more digging and there is an official Nextcloud image based on Debian as well, and installing the NVIDIA drivers on Debian is completely doable. 

Would it be possible to go the route many others have, and make a separate NVIDIA image? This way people who don't need NVIDIA support don't have a needlessly large Docker image. And those who do use NVIDIA can not only get hardware transcoding but also TensorFlow, CUDA, & cuDNN for use with the Recognize app. 

To keep size down as much as possible, my thought would be to have an env variable like "ENABLE_CUDA", if its set to true, then on the start of the container a script checks to see if the libs already exist, and if not then they get downloaded and installed to a persistent directory. Same thing for ffmpeg, it doesn't take long to compile, and I'd gladly wait 20-30min one time if it meant I didn't need to run NC in a VM with a dedicated GPU.

**Reply by szaimen** [reply]
Your best bet is to request the compatibility of ffmpeg with nvenc in upstream alpine. See https://github.com/nextcloud/all-in-one/discussions/1536#discussioncomment-4422566

**Reply by szaimen** [reply]
As for recognize, this sounds like the best approach to me: https://github.com/nextcloud/recognize/issues/73

**Reply by Zk2u** [reply]
Would it be possible to get GPU support in AIO as an opt in? The recognize issue doesn't appear to be making much progress. I suppose it could be retrofitted in by modifying the container but i don't want to tamper with the existing installation.

**Reply by szaimen** [reply]
Hi, regarding CUDA, see https://github.com/nextcloud/all-in-one/pull/5132#issuecomment-2556900348

**Comment by shperrung** [comment]
Excuse me for dump question.
Is v4.1.0 beta release about HW nvenc/cuda support in ffmpeg?
I tried to build ffmpeg with cuda/nvenc libs in docker using JRottenberg's approaches but stacked in missing "libnppc". Unfortunatelly it is a part of nvidia-cuda-toolkit. It is very diffcult for me to get it into te same image with nextloud for successful build  and work ffmpeg.

**Reply by szaimen** [reply]
Hi, see the discussion in https://github.com/nextcloud/all-in-one/discussions/1536

**Reply by szaimen** [reply]
Hi, see https://github.com/nextcloud/all-in-one/pull/5132#issuecomment-2556900348

**Comment by chiva** [comment]
Is `libva-utils` required?

I can't install it directly within the Nextcloud `NEXTCLOUD_ADDITIONAL_APKS`
Which I have set to this at the moment to use QuickSync within Nextcloud:
`imagemagick libva-utils libva-vdpau-driver libva-intel-driver intel-media-driver mesa-va-gallium`

```
/var/www/docker-aio # apk add libva-utils
fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/community/x86_64/APKINDEX.tar.gz
ERROR: unable to select packages:
  libva-utils (no such package):
    required by: world[libva-utils]
```

Checking the Alpine repo, `libva-utils` is unavailable in the `main` repository, only in the `testing` one.
https://pkgs.alpinelinux.org/packages?name=libva-utils&branch=edge&repo=&arch=&maintainer=

So you would need to do:
`apk add libva-utils --repository=http://dl-cdn.alpinelinux.org/alpine/edge/testing/`


------------
Also, is the following output expected to be correct?
I see some 404s and failures when testing VAAPI, which I'd expect to work having an Intel CPU 🤔
 
```
# docker exec -it --user www-data nextcloud-aio-nextcloud php occ memories:video-setup

This command will help you setup video streaming
It will check if ffmpeg and ffprobe are installed and go-vod (the transcoder) works
If you are running go-vod externally, feature detection will not work properly
In that case you will need to manually adjust the configuration file (add --print-config)
For details on hardware support: https://github.com/pulsejet/memories/wiki/HW-Transcoding

Press enter to continue or CTRL+C to abort

ffmpeg is installed
ffprobe is installed
Checking for go-vod binary
Trying go-vod from /var/www/html/custom_apps/memories/exiftool-bin/go-vod-amd64
go-vod is installed!

You can use transcoding and HLS streaming
This is recommended for better performance, but has implications if
you are using external storage or run Nextcloud on a slow system.

Read the following documentation carefully before continuing:
https://github.com/pulsejet/memories/wiki/Configuration

Do you want to enable transcoding and HLS? [Y/n]


Starting ffmpeg feature detection
This may take a while. Please be patient

Downloading test video file ... OK

Attempting to start go-vod ... OK
go-vod logs will be stored at: /tmp/go-vod/oczbevt4qe6s.log

Testing transcoding with CPU ... FAIL
CPU transcoding failed with error HTTP status: 404
Check the log file of go-vod for more details (/tmp/go-vod/oczbevt4qe6s.log)

Checking for VAAPI acceleration (/dev/dri/renderD128) ... OK
Checking for permissions on /dev/dri/renderD128 ... OK
Testing transcoding with VAAPI ... FAIL
VAAPI transcoding failed with error HTTP status: 404
Check the log file of go-vod for more details (/tmp/go-vod/oczbevt4qe6s.log)
Testing transcoding with VAAPI (low_power) ... FAIL
VAAPI (low_power) transcoding failed with error HTTP status: 404
Check the log file of go-vod for more details (/tmp/go-vod/oczbevt4qe6s.log)

Checking for NVIDIA acceleration with NVENC
Testing transcoding with NVENC (scale_npp) ... FAIL
Testing transcoding with NVENC (scale_cuda) ... FAIL
NVENC does not seem to be available

Feature detection completed

Transcoding and HLS are now enabled! Monitor the log file for any errors
You should restart the server for changes to take effect
```

If I recall, in a previous install test I had those same errors but they went away once I installed all the packages with apk.

cc @relink2013 @szaimen

**Reply by Zoey2936** [reply]
please request [here](https://gitlab.alpinelinux.org/alpine/aports/-/issues), if the package could be moved outside testing

**Reply by chiva** [reply]
thanks @Zoey2936 !
Do we know if that package is really required?

I have installed it manually through the console, but seeing some errors on the output of `memories:video-setup`
Are they expected? Am I missing any other package? Did I missed some command or restart?

**Comment by szaimen** [comment, marked as answer]
Nvidia gpu passthrough is now being worked on in https://github.com/nextcloud/all-in-one/pull/5132

**Reply by szaimen** [reply]
This is now released with v10.2.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel
